rai/node/rpc.cpp

```diff
@@ -4228,23 +4228,40 @@ void rai::rpc_handler::work_generate ()
 	if (rpc.config.enable_control)
 	{
 		std::string hash_text (request.get<std::string> ("hash"));
+	    bool use_peers (request.get_optional<bool> ("use_peers") == true);
 		rai::block_hash hash;
 		auto error (hash.decode_hex (hash_text));
 		if (!error)
 		{
 			auto rpc_l (shared_from_this ());
-			node.work.generate (hash, [rpc_l](boost::optional<uint64_t> const & work_a) {
-				if (work_a)
-				{
-					boost::property_tree::ptree response_l;
-					response_l.put ("work", rai::to_string_hex (work_a.value ()));
-					rpc_l->response (response_l);
-				}
-				else
-				{
-					error_response (rpc_l->response, "Cancelled");
-				}
-			});
+			if (!use_peers) {
+				node.work.generate (hash, [rpc_l](boost::optional<uint64_t> const & work_a) {
+					if (work_a)
+					{
+						boost::property_tree::ptree response_l;
+						response_l.put ("work", rai::to_string_hex (work_a.value ()));
+						rpc_l->response (response_l);
+					}
+					else
+					{
+						error_response (rpc_l->response, "Cancelled");
+					}
+				});
+			} else {
+				node.generate_work (hash, [rpc_l](boost::optional<uint64_t> const & work_a) {
+					if (work_a)
+					{
+						boost::property_tree::ptree response_l;
+						response_l.put ("work", rai::to_string_hex (work_a.value ()));
+						rpc_l->response (response_l);
+					}
+					else
+					{
+						error_response (rpc_l->response, "Cancelled");
+					}
+				});
+			}
+
```

Comment:
Formatting issues aside, there seems to be some code duplication going here. I'd recommend putting the inner if/else after the outer which should merely stuff the `to_string_hex (...)` result into variable.

Reply:
Hm yeah I didn't like to do it that way. But I am not that much into cpp and don't really understand how the functor-like coding works. I mean best would be if you could save this function to some value (callback for example):

```
callback = [rpc_l](boost::optional<uint64_t> const & work_a) {
	if (work_a)
	{
		boost::property_tree::ptree response_l;
		response_l.put ("work", rai::to_string_hex (work_a.value ()));
		rpc_l->response (response_l);
	}
	else
	{
		error_response (rpc_l->response, "Cancelled");
	}
}
```

And then having the function passed to the generate functions:

```
if (!use_peers) {
	node.work.generate (hash, callback);
} else {
	node.generate_work (hash, callback);
}
```

No idea how to do that in cpp though.

Reply:
Yeah, that'll work, just put `auto` in front of the callback definition and you should be good to go (do test it though).

Reply:
Alright worked exactly like that. Feel free to check it out again.